internal/incremental/emitfileshandler.go

```diff
@@ -0,0 +1,280 @@
+					// With d.ts diagnostics they are also part of the signature so emitSignature will be different from it since its just hash of d.ts
+					if len(data.Diagnostics) == 0 {
+						emitSignature = signature
+					}
+					if signature != info.version { // Update it
+						h.signatures.Store(options.TargetSourceFile.Path(), signature)
+					}
+				}
+
+				// Store d.ts emit hash so later can be compared to check if d.ts has changed.
+				// Currently we do this only for composite projects since these are the only projects that can be referenced by other projects
+				// and would need their d.ts change time in --build mode
+				if h.skipDtsOutputOfComposite(options.TargetSourceFile, fileName, text, data, emitSignature) {
+					return nil
+				}
+			}
+
+			if options.WriteFile != nil {
+				return options.WriteFile(fileName, text, writeByteOrderMark, data)
+			}
+			return h.program.program.Host().FS().WriteFile(fileName, text, writeByteOrderMark)
+		},
+	}
+}
+
+// Compare to existing computed signature and store it or handle the changes in d.ts map option from before
+// returning undefined means that, we dont need to emit this d.ts file since its contents didnt change
+func (h *emitFilesHandler) skipDtsOutputOfComposite(file *ast.SourceFile, outputFileName string, text string, data *compiler.WriteFileData, newSignature string) bool {
+	if !h.program.snapshot.options.Composite.IsTrue() {
+		return false
+	}
+	var oldSignature string
+	oldSignatureFormat, ok := h.program.snapshot.emitSignatures[file.Path()]
+	if ok {
+		if oldSignatureFormat.signature != "" {
+			oldSignature = oldSignatureFormat.signature
+		} else {
+			oldSignature = oldSignatureFormat.signatureWithDifferentOptions[0]
+		}
+	}
+	if newSignature == "" {
+		newSignature = h.program.snapshot.computeHash(getTextHandlingSourceMapForSignature(text, data))
+	}
+	// Dont write dts files if they didn't change
+	if newSignature == oldSignature {
+		// If the signature was encoded as string the dts map options match so nothing to do
+		if oldSignatureFormat != nil && oldSignatureFormat.signature == oldSignature {
+			data.SkippedDtsWrite = true
+			return true
+		} else {
+			// Mark as differsOnlyInMap so that --build can reverse the timestamp so that
+			// the downstream projects dont detect this as change in d.ts file
+			data.DiffersOnlyInMap = true
+		}
+	} else {
+		h.latestChangedDtsFiles.Add(outputFileName)
+	}
+	h.emitSignatures.Store(file.Path(), &emitSignature{signature: newSignature})
+	return false
+}
+
+func (h *emitFilesHandler) updateSnapshot() []*compiler.EmitResult {
+	h.signatures.Range(func(file tspath.Path, signature string) bool {
+		info := h.program.snapshot.fileInfos[file]
+		info.signature = signature
+		if h.program.updatedSignatureKinds != nil {
+			h.program.updatedSignatureKinds.Store(file, SignatureUpdateKindStoredAtEmit)
+		}
+		h.program.snapshot.buildInfoEmitPending = true
```

Comment:
I haven't gotten there yet, but it strikes me as odd that we are atomically mutating one thing at the same time as we are not atomically mutating another thing, when both are sort of "on" the same Program... I can't really put my finger on it but it gives me the bad feeling of the model not being consistent and free of logical races.